### DISCUSSION TITLE
When will the SPI transaction reset up after the otPlatSpiSlavePrepareTransaction returns busy.

### DISCUSSION THREAD

**Discussion by zwx1995esp** [openthread/openthread#9704] [Q&A] [answered]
I encountered a bug when using the NCP to communicate with the RCP using SPI spinel, after some investigations, I found that the transaction of `Tx done` notifying to master is busy,  this packet will never reach to the platform layer, seems that the resend logic(or mechanism) does not work. 

In the [NCP code](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_spi.cpp#L297), when the current SPI transaction is busy, there is a comment showing that `We will get the transaction set up properly when the current transaction is completed.` It shows that the openthread stack will restart the transcation properly, but I did not see any logic or mechanism of that... I create this discussion here to find some help. Is there someone who can illustrate these resending logics? Thanks.

**Comment by abtink** [comment]
@zwx1995esp It is been a while since i last looked at `NcpSpi` but will try to provide some info:

```c++
  mTxState = kTxStateSending;

  error = otPlatSpiSlavePrepareTransaction(mSendFrame, mSendFrameLength, nullptr, 0, /* aRequestTrans */ true);

    if (error == OT_ERROR_BUSY)
    {
        // Being busy is OK. We will get the transaction set up
        // properly when the current transaction is completed.
        error = OT_ERROR_NONE;
    }
```
- Note that we set ` mTxState = kTxStateSending` before this call.
- If `otPlatSpiSlavePrepareTransaction()` returns `OT_ERROR_BUSY` an SPI transaction is active (in process).
- When the transaction is done platform must call the callback `otPlatSpiSlaveTransactionCompleteCallback` to indicate this.
- In `NcpSpi` this callback maps to `NcpSpi::SpiTransactionComplete()` method.
- This method then prepares the next transaction based on current state
-  For example, it checks and if `mTxState == kTxStateSending` it uses `mSendFrame` for next transacation.

```c++
bool NcpSpi::SpiTransactionComplete(...
{
   ...

   // Determine the input and output frames to prepare a new transaction.

    if (mResetFlag && (aTransLen > 0) && (aOutputLen > 0))
    {
        mResetFlag = false;
        sendFrame.SetHeaderFlagByte(/*aResetFlag */ false);
        SpiFrame(mEmptySendFrameFullAccept).SetHeaderFlagByte(/*aResetFlag */ false);
        SpiFrame(mEmptySendFrameZeroAccept).SetHeaderFlagByte(/*aResetFlag */ false);
    }

    if (mTxState == kTxStateSending)
    {
        aOutputBuf = mSendFrame;
        aOutputLen = mSendFrameLength;
    }
    else
    {
        aOutputBuf = mHandlingRxFrame ? mEmptySendFrameZeroAccept : mEmptySendFrameFullAccept;
        aOutputLen = kSpiHeaderSize;
    }

**Reply by zwx1995esp** [reply]
Thanks a lot!

**Reply by zwx1995esp** [reply]
Hi @abtink  but for the logic in function `NcpSpi::SpiTransactionComplete`, 
```
    if (mTxState == kTxStateSending)
    {
        uint16_t txDataLen = outputFrame.GetHeaderDataLen();
        // Frame transmission is successful if master indicates
        // in the input frame that it could accept the frame
        // length that was exchanged, i.e., the "data len" in
        // the output frame is smaller than or equal to "accept
        // len" in the received input frame from master.
        if ((txDataLen > 0) && (txDataLen <= transDataLen) && (txDataLen <= inputFrame.GetHeaderAcceptLen()))
        {
             mTxState      = kTxStateHandlingSendDone;
             shouldProcess = true;
        }
```
It shows that if the current transcation has successfully done, and if no errors with the checking txDataLen(this check might confirm that the master did receive the transmited frames), then the state will be set to `kTxStateHandlingSendDone`. So can I believe that the checking `mTxState == kTxStateSending` after the key words `exit` is only to check if current transcation is successful or not, if not, then it will retry again.

Anythings wrong in my thought?

**Reply by zwx1995esp** [reply]
Also I add a global variable `flag` to indicate if current transcation is busy.  Before the txDataLen check logic, I check if the flag is true first:
```
    if (mTxState == kTxStateSending)
    {
        if (flag) {
             // Do nothing.
        } else {
            uint16_t txDataLen = outputFrame.GetHeaderDataLen();
            // Frame transmission is successful if master indicates
            // in the input frame that it could accept the frame
            // length that was exchanged, i.e., the "data len" in
            // the output frame is smaller than or equal to "accept
            // len" in the received input frame from master.
            if ((txDataLen > 0) && (txDataLen <= transDataLen) && (txDataLen <= inputFrame.GetHeaderAcceptLen()))
            {
                 mTxState      = kTxStateHandlingSendDone;
                 shouldProcess = true;
            }
        }
        flag = 0;
```
and set this flag in function `PrepareNextSpiSendFrame` when `otPlatSpiSlavePrepareTransaction` returns `OT_ERROR_BUSY`:

```
void NcpSpi::PrepareNextSpiSendFrame(void)
{
    ......
    error = otPlatSpiSlavePrepareTransaction(mSendFrame, mSendFrameLength, nullptr, 0, /* aRequestTrans */ true);
    if (error == OT_ERROR_BUSY) 
    {
        // Being busy is OK. We will get the transaction set up
        // frame while keeping the input frame unchanged
        error = OT_ERROR_NONE;
        flag = 1;
    }
    ......
````
(These changes are only for debug, not a reasonable fix if this was a issue. I do not consider the corner case when there is a pending transcation but current transcation is failed....)
After this change, the resend frame reaches to the platform. I also test it with a `tcp iperf` stress test with average `23.96 Kbits/sec` throughput running 2000 seconds, only see `tx timeout error` and `RCP recover` twice(Previously, with the issue, the `tx timeout error` and `RCP recover` appear every 4~10 seconeds).

So,do you think there are some bugs for the retransmition logic?

**Reply by abtink** [reply, marked as answer]
Hi @zwx1995esp 

I may be wrong but I think the code should handle such cases correctly. Let me describe my thoughts/reasons:

About this part of code (in `SpiTransactionComplete()`) which checks for `kTxStateSending` we also check `(txDataLen > 0)` ensuring that the transaction that completed did indeed contain tx data.

```c++
   if (mTxState == kTxStateSending)
    {
        uint16_t txDataLen = outputFrame.GetHeaderDataLen();

        // Frame transmission is successful if master indicates
        // in the input frame that it could accept the frame
        // length that was exchanged, i.e., the "data len" in
        // the output frame is smaller than or equal to "accept
        // len" in the received input frame from master.

        if ((txDataLen > 0) && (txDataLen <= transDataLen) && (txDataLen <= inputFrame.GetHeaderAcceptLen()))
        {
            mTxState      = kTxStateHandlingSendDone;
            shouldProcess = true;
        }
    }
```


The TX frames are processed sequentially and we prepare the next tx frame only after the previous one is successfully sent. This is tracked by the `mTxState` changes. `PrepareNextSpiSendFrame()` is only called from `PrepareTxFrame()` when we are in `kTxStateIdle`:

```c++
void NcpSpi::PrepareTxFrame(void)
{
    switch (mTxState)
    {
    case kTxStateHandlingSendDone:
        mTxState = kTxStateIdle;

        OT_FALL_THROUGH;
        // to next case to prepare the next frame (if any).

    case kTxStateIdle:
        PrepareNextSpiSendFrame();
        break;

    case kTxStateSending:
        // The next frame in queue (if any) will be prepared when the
        // current frame is successfully sent and this task is posted
        // again from the `SpiTransactionComplete()` callback.
        break;
    }
}
```
 - If previous tx was success and we are in `kTxStateHandlingSendDone` and we then transition to `kTxStateIdle`.
 - Or if we are already in `kTxStateIdel`.

In both cases, the last prepared transaction should have empty TX data.
- So when we are in `PrepareNextSpiSendFrame()` the current active transition should have empty TX data.
- And if the call `otPlatSpiSlavePrepareTransaction()` from `PrepareNextSpiSendFrame()` does return `BUSY` error the transaction that would be ongoing should have empty TX data.
- Which means when `SpiTransactionComplete()` is called, the check `(txDataLen > 0)` should be `false` and ensure that we would not consider the tx finished (stay in `kTxStateSending` state).
- Which would then properly prepare the next transaction with the TX data to send.

----

One maybe related topic on platform SPI `otPlatSpi` implementation.

We expect/require that after a transaction is completed and `otPlatSpiSlaveTransactionCompleteCallback` is invoked, the previously prepared transaction must no longer be used. The platform SPI implementation must clear both input/output buffers. I think if this is not done by platform implementation (and if `otPlatSpiSlaveTransactionCompleteCallback` is called with some delay after SPI transaction is completed) , we may see bad/weird behaviors. 
```c++
/**
 * Indicates that a SPI transaction has completed with the given length. The data written to the slave has been written
 * to the pointer indicated by the `aInputBuf` argument to the previous call to `otPlatSpiSlavePrepareTransaction()`.
 *
 * Once this function is called, `otPlatSpiSlavePrepareTransaction()` is invalid and must be called again for the next
 * transaction to be valid.
 *
 * Note that this function is always called at the end of a transaction, even if `otPlatSpiSlavePrepareTransaction()`
 * has not yet been called. In such cases, `aOutputBufLen` and `aInputBufLen` will be zero.
 *
```

**Reply by zwx1995esp** [reply]
Hi, Abtink, I think I get your points about what you said `the last prepared transaction should have empty TX data`:

The slave can not start a SPI transcation, if the slave wants to send packets to master, it should notify the master first then master starts the transcation. After slave transcation is done, then a new enpty frame will be added to send to platform, and platform needs to configure it to SPI peripheral. The empty frame is used for exchanging data from master(becasue there is a valid pattern in the SPI frame header). So if `otPlatSpiSlavePrepareTransaction` return busy on slave, it indicates that the current SPI HW is transmiting some packets from master to slave, and the SPI slave must transmit the empty frames. And on slave end, OT stack will not transmit two packets at the same time.

Is this time right? If right, I think that the logic seems reasonable and I can also understand how does the re-transmit logic work.

**Reply by abtink** [reply]
Yes. Exactly. The call to `otPlatSpiSlavePrepareTransaction()` returning BUSY can only happen when we are in `TxStateIdle` which means we don't have any data to send to controller/master.

**Reply by zwx1995esp** [reply]
Hi Abtink, thanks a lot, seems some bugs in our own platform code. I will investigate it a bit more.
